### ISSUE TITLE
PDF-Hul produces Invalid Page Dictionary for PDF's with VP dictionaries

### ISSUE THREAD

**Issue by thorsted** [openpreserve/jhove#662]
Any PDF which uses a [ViewPort dictionary](https://opensource.adobe.com/dc-acrobat-sdk-docs/pdfstandards/PDF32000_2008.pdf#page=487) including the newer [Geo Supplement](https://www.adobe.com/content/dam/acom/en/devnet/acrobat/pdfs/adobe_supplement_iso32000.pdf#page=49) throws the [PDF-HUL-117](https://github.com/openpreserve/jhove/wiki/PDF-hul-Messages-2#pdf-hul-117) error. 

May be related to the parsing of the Measurement key.

[AutoCAD-Viewport-Sample1.pdf](https://github.com/openpreserve/jhove/files/6063282/AutoCAD-Viewport-Sample1.pdf)

**Comment by bdoubrov** [comment]
Indeed, current way of parsing measurement format arrays is incorrect and will be fixed. The only remaining question is how to present this measurement data in the report. For example, there might be multiple measurement units in X (or Y) direction. In PDF they would be defined as (see Table 262 in PDF 1.7 spec):
```
<<
... 
/X  [ << /U (inch) /C 0.0254 >> << /U (mile) /C 1609.34>> ]
...
>> 
```
We can put them in the report as is:
```
-- X
---- U inch
---- C 0.0254
---- U mile
---- C 1609.34
```
or make the key names more user friendly. Like:
```
-- X
---- Name "inch"
---- Coefficient 0.0254
---- Name "mile"
---- Coefficient 1609.34
```

**Comment by thorsted** [comment]
I have found in most of my samples the units and ratio are often not populated, Do they need to be?
``` 
/Measure <<
          /A [ <<
                /C 1
                /U ( )
              >> ]
          /D [ <<
                /C 1
                /U ( )
              >> ]
          /R ( )
          /Subtype /RL
          /Type /Measure
          /X [ <<
                /C .00231
                /U ( )
              >> ]
        >>
```

**Comment by bdoubrov** [comment]
Both /U and /C entries are required (see Table 263, PDF 1.7 spec). /U entry is defined as:
> U - (Required) A text string specifying a label for displaying the units represented by this dictionary in a user interface; the label should use a universally recognized abbreviation.

The example from PDF 1.7 specification suggests the following use:
```
<</Type /Measure
/Subtype /RL
/R (1in = 0.1 mi)
/X [ <</U (mi) % x offset represented in miles
/C .00139 % Conversion from user space units to miles
/D 100000
]
/D[<< /U (mi) /C 1 >>% Distance: initial unit is miles; no conversion needed
<< /U (ft) /C 5280 >>% Conversion from miles to feet
<< /U (in) /C 12 % Conversion from feet to inches
/F /F /D 8 >>% Fractions of inches rounded to nearest 1/8
]
/A [<</U (acres) % Area: measured in acres
/C 640 >>% Conversion from square miles to acres
]
>>
```

However, an empty /U value is not forbidden. And this is what the PDF creators might use in cases when the scaling coefficient is not linked to any standard unit.

**Comment by thorsted** [comment]
1.26RC1 PDF-Hul is still using version 1.12.2, so can't test.

**Comment by carlwilson** [comment]
This one I suspected might be a mistake on my part. Now I'm aware there's a packaging issue, which likely has a similar cause, I'll still double check, but believe it's fixed.

**Comment by thorsted** [comment]
> This one I suspected might be a mistake on my part. Now I'm aware there's a packaging issue, which likely has a similar cause, I'll still double check, but believe it's fixed.

Just checked PDF-Hul 1.12.3 and this sample PDF is still not well formed.

**Comment by carlwilson** [comment]
I'll look into this and get back to you this week. I think I may have been hard of understanding on the issue, let's see.
